File: Solutions/337_House_Robber_III/337_House_Robber_III.go
//Second fastest
func rob(root *TreeNode) int {
    ans0, ans1 := helper(root)
    return max(ans0, ans1)
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func helper(root *TreeNode) (int, int) {
    if root == nil {
        return 0, 0
    }
    left0, left1 := helper(root.Left)
    right0, right1 := helper(root.Right)
    self0, self1 := max(left0, left1) + max(right0, right1), root.Val + left0 + right0
    return self0, self1
}

//Fastest
func rob(root *TreeNode) int {
    rob, not := myRob(root)
    return max(rob, not)
}

func myRob(root *TreeNode)(rob, not int) {
    if root == nil {
        return 0, 0
    }
    leftR, leftN := myRob(root.Left)
    rightR, rightN := myRob(root.Right)
    
    return root.Val + leftN + rightN, max(leftN, leftR) + max(rightR, rightN)
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: Solutions/337_House_Robber_III/337_House_Robber_III.c
//Fastest
#define max(x,y) (x>y?x:y)

typedef struct TreeNode* Node;
struct robber{
    int if_rob;
    int if_not_rob;
};
typedef struct robber ROB;
struct robber helper(Node root){
    ROB local={0,0};
    if(root){
        ROB left = helper(root->left);
        ROB right= helper(root->right);
        local.if_not_rob=max(left.if_rob,left.if_not_rob)+max(right.if_rob,right.if_not_rob);
        local.if_rob=root->val+left.if_not_rob+right.if_not_rob;
        return local;
        
    }else{
        return local;
    }
    
}

int rob(struct TreeNode* root){
    ROB money=helper(root);
    return max(money.if_rob,money.if_not_rob);
    
}